src/app/shared/data-input/data-input.component.html

```diff
@@ -0,0 +1,49 @@
+<ng-container *ngIf="constraint; else textDataInput"
+              [ngSwitch]="constraint.type">
+  <boolean-data-input *ngSwitchCase="'Boolean'"
+                      [value]="value"
+                      (valueChange)="onValueChange($event)"
+                      (save)="save.emit($event)">
+  </boolean-data-input>
+
+  <datetime-data-input *ngSwitchCase="'DateTime'"
+                       [constraintConfig]="constraint.config"
+                       [focus]="focus"
+                       [readonly]="readonly"
+                       [value]="value"
+                       (valueChange)="onValueChange($event)"
+                       (save)="save.emit($event)"
+                       (cancel)="cancel.emit()">
+  </datetime-data-input>
+
+  <number-data-input *ngSwitchCase="'Number'"
+                     [focus]="focus"
+                     [readonly]="readonly"
+                     [constraintConfig]="constraint.config"
```

Comment:
`contraint?.config` ... also above

Reply:
There is an `if` condition on the first line of this template so `constraint` cannot be null here.